### PR TITLE
sql: Update role membership for PostgreSQL v16

### PR DESCRIPTION
**Pull request by jkosh44**
Previously, role membership was modeled after the implementation of role membership in PostgreSQL v15. Specifically the grantor was set as the role ID of the session executing the `GRANT` statement. However, PostgreSQL v16, which has not been released yet as of this commit, updated role membership to be more consistent with the SQL standard and with object privileges. Specifically the grantor could only be set to someone with ADMIN OPTION on the role being granted or the bootstrap superuser.

This commit set's the grantor of all role membership to the mz_system role. We haven't implemented ADMIN OPTION yet, and the 'mz_system' role is our version of the bootstrap superuser.

The PostgreSQL commit that made this change can be found here: https://github.com/postgres/postgres/commit/ce6b672e4455820a0348214be0da1a024c3f619f

Part of #11579

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/sql/src/plan.rs [hunk 1]
```diff
@@ -807,6 +807,8 @@ pub struct RevokeRolePlan {
     pub role_id: RoleId,
     /// The roles that will be removed from `role_id`.
     pub member_ids: Vec<RoleId>,
+    /// The role that granted the membership.
```
Comment (mjibson) [inline src/sql/src/plan.rs:810]:
```suggestion
    /// The role that revoked the membership.
```

File: src/sql/src/plan/statement/ddl.rs [hunk 2]
```diff
@@ -4213,7 +4214,16 @@ pub fn plan_grant_role(
         member_names,
     }: GrantRoleStatement<Aug>,
 ) -> Result<Plan, PlanError> {
-    let grantor_id = scx.catalog.active_role_id().clone();
+    // In PostgreSQL, the grantor must either be a role with ADMIN OPTION on the role being granted,
+    // or the bootstrap superuser. We do not have ADMIN OPTION implemented and 'mz_system' is our
+    // equivalent of the bootstrap superuser. Therefore the grantor is always 'mz_system'.
+    // For more details see:
+    // https://github.com/postgres/postgres/blob/064eb89e83ea0f59426c92906329f1e6c423dfa4/src/backend/commands/user.c#L2180-L2238
+    let grantor_id = scx
```
Comment (mjibson) [inline src/sql/src/plan/statement/ddl.rs:4222]: Will this make the audit log useless for looking up which human typed grant or revoke? If so, we should delay this PR until ADMIN OPTION is implemented somehow. Can ADMIN OPTION be implied for superusers?

Comment (jkosh44) [reply src/sql/src/plan/statement/ddl.rs:4222]: > Will this make the audit log useless for looking up which human typed grant or revoke?

Pretty much yes.

> If so, we should delay this PR until ADMIN OPTION is implemented somehow.

Even when `ADMIN OPTION` is implemented, the grantor role does not always reflect the role that executed the `GRANT` statement. It reflects the role chosen as the best grantor.

> Can ADMIN OPTION be implied for superusers.

Yes in the sense that superusers can always execute `GRANT`. No in the sense that the superuser will not be chosen as the grantor unless they have `ADMIN OPTION` explicitly.

Comment (jkosh44) [reply src/sql/src/plan/statement/ddl.rs:4222]: Maybe the audit log entry should contain an additional field of something like "session role membership".

Comment (jkosh44) [reply src/sql/src/plan/statement/ddl.rs:4222]: I updated the audit log to also contain the role of the session that execute the command.

If you're curious the pg mailing thread that discusses this change can be found here: https://www.postgresql.org/message-id/flat/1102925.1654199458%40sss.pgh.pa.us#505bfe4947d7a86ee07d5fdad85f34d1

It's very long and mostly irrelevant to this change (I haven't finished reading the whole thing yet). Scroll down to where the email subject changes from "CREATEROLE and role ownership hierarchies" to "pg_auth_members.grantor is bunk" for the relevant emails.

**Review by def- — APPROVED** [review]
What is our guidance on compatibility with Postgres? Should it always be with the latest version, if that is a breaking change?

**Comment by jkosh44** [comment]
> What is our guidance on compatibility with Postgres? Should it always be with the latest version, if that is a breaking change?

In general we try and target the most recent version of PostgreSQL, though I'm not sure if that's written down anywhere. For breaking changes we can address them on a case-by-case basis. However, PostgreSQL is also reluctant to make breaking changes, so if they're making the change then it's probably an important change.

**Review by mjibson — APPROVED** [review]
I haven't read the postgres thread to understand their reasoning, but I trust them. The `executed_by` fields in the audit log seem good. Question: if the grantor fields were bunk before this change, are the `executed_by` fields also bunk?

**Comment by jkosh44** [comment]
> I haven't read the postgres thread to understand their reasoning, but I trust them. The `executed_by` fields in the audit log seem good. Question: if the grantor fields were bunk before this change, are the `executed_by` fields also bunk?

I haven't read every single message, but my understanding of "bunk" in this context is that it means:
- Not compliant with the SQL spec.
- May refer to a role that's been deleted.
- Does not form a tree with the builtin superuser as the root. The tree allows you to trace the lineage of all grants.
- Is inconsistent with object privilege grantors.
- Doesn't allow duplicate privileges with different grantors.

None of these really apply to `executed_by` which just allows us to audit who actually typed the command.

**Pull request merged**